File: godot-java-bridge/src/main/java/io/github/kingg22/godot/internal/bridge/ClassDBBridge.java
package io.github.kingg22.godot.internal.bridge;

import io.github.kingg22.godot.api.GodotClass;
import io.github.kingg22.godot.internal.ffm.GDExtensionCallError;
import io.github.kingg22.godot.internal.ffm.GDExtensionClassCreateInstance2;
import io.github.kingg22.godot.internal.ffm.GDExtensionClassCreationInfo5;
import io.github.kingg22.godot.internal.ffm.GDExtensionClassFreeInstance;
import io.github.kingg22.godot.internal.ffm.GDExtensionClassMethodCall;
import io.github.kingg22.godot.internal.ffm.GDExtensionClassMethodInfo;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Supplier;

import static java.util.Objects.requireNonNull;

/** High-level ClassDB registration and instance dispatch. */
public final class ClassDBBridge {
    private static final int CALL_OK = 0;

    private final GodotFFI ffi;
    private final StringNameCache stringNames;
    private final Arena arena;
    private final AtomicLong classIdGenerator = new AtomicLong(1);
    private final AtomicLong methodIdGenerator = new AtomicLong(1);
    private final Map<Long, ClassEntry> classes = new ConcurrentHashMap<>();
    private final Map<Long, MethodHandler> methods = new ConcurrentHashMap<>();
    private final Map<Long, InstanceHandle> instances = new ConcurrentHashMap<>();

    private final MemorySegment methodCallStub;
    private final MemorySegment createInstanceStub;
    private final MemorySegment freeInstanceStub;

    ClassDBBridge(final GodotFFI ffi, final StringNameCache stringNames, final Arena arena) {
        this.ffi = ffi;
        this.stringNames = stringNames;
        this.arena = arena;
        this.methodCallStub = GDExtensionClassMethodCall.allocate(this::onMethodCall, arena);
        this.createInstanceStub = GDExtensionClassCreateInstance2.allocate(this::onCreateInstance, arena);
        this.freeInstanceStub = GDExtensionClassFreeInstance.allocate(this::onFreeInstance, arena);
    }

    public void registerClass(final ClassDefinition definition) {
        final long classId = classIdGenerator.getAndIncrement();
        final var classUserdata = arena.allocate(ValueLayout.JAVA_LONG);
        classUserdata.set(ValueLayout.JAVA_LONG, 0, classId);

        final var className = stringNames.get(definition.className());
        final var parentName = stringNames.get(definition.parentClassName());

        final var creationInfo = GDExtensionClassCreationInfo5.create(
                false,
                false,
                true,
                createInstanceStub,
                freeInstanceStub,
                "",
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                null,
                classUserdata);

        classes.put(classId, new ClassEntry(definition, className, parentName, classUserdata));
        ffi.classdbRegisterExtensionClass5(className, parentName, creationInfo);
    }

    public void registerMethod(final String className, final String methodName, final MethodHandler handler) {
        final long methodId = methodIdGenerator.getAndIncrement();
        final var methodUserdata = arena.allocate(ValueLayout.JAVA_LONG);
        methodUserdata.set(ValueLayout.JAVA_LONG, 0, methodId);
        methods.put(methodId, handler);

        final var methodInfo = arena.allocate(GDExtensionClassMethodInfo.layout());
        GDExtensionClassMethodInfo.name(methodInfo, stringNames.get(methodName));
        GDExtensionClassMethodInfo.method_userdata(methodInfo, methodUserdata);
        GDExtensionClassMethodInfo.call_func(methodInfo, methodCallStub);
        GDExtensionClassMethodInfo.ptrcall_func(methodInfo, MemorySegment.NULL);
        GDExtensionClassMethodInfo.method_flags(methodInfo, 0);
        GDExtensionClassMethodInfo.has_return_value(methodInfo, (byte) 0);
        GDExtensionClassMethodInfo.return_value_info(methodInfo, MemorySegment.NULL);
        GDExtensionClassMethodInfo.return_value_metadata(methodInfo, 0);
        GDExtensionClassMethodInfo.argument_count(methodInfo, 0);
        GDExtensionClassMethodInfo.arguments_info(methodInfo, MemorySegment.NULL);
        GDExtensionClassMethodInfo.arguments_metadata(methodInfo, MemorySegment.NULL);
        GDExtensionClassMethodInfo.default_argument_count(methodInfo, 0);
        GDExtensionClassMethodInfo.default_arguments(methodInfo, MemorySegment.NULL);

        ffi.classdbRegisterExtensionClassMethod(stringNames.get(className), methodInfo);
    }

    private MemorySegment onCreateInstance(final MemorySegment classUserdata, final byte notifyPostInitialize) {
        final long classId =
                classUserdata.reinterpret(ValueLayout.JAVA_LONG.byteSize()).get(ValueLayout.JAVA_LONG, 0);
        final var entry = classes.get(classId);
        if (entry == null) {
            return MemorySegment.NULL;
        }

        final var objectPtr = ffi.classdbConstructObject2(entry.className);
        if (objectPtr == MemorySegment.NULL) {
            return MemorySegment.NULL;
        }

        final var handle = InstanceHandle.create(entry.definition.factory());
        instances.put(handle.address, handle);
        ffi.objectSetInstance(objectPtr, entry.className, handle.dataPointer);
        return objectPtr;
    }

    private void onFreeInstance(final MemorySegment classUserdata, final MemorySegment instancePtr) {
        final var address = instancePtr.address();
        final var handle = instances.remove(address);
        if (handle != null) {
            handle.close();
        }
    }

    private void onMethodCall(
            final MemorySegment methodUserdata,
            final MemorySegment instancePtr,
            final MemorySegment args,
            final long argCount,
            final MemorySegment returnValue,
            final MemorySegment error) {
        final long methodId =
                methodUserdata.reinterpret(ValueLayout.JAVA_LONG.byteSize()).get(ValueLayout.JAVA_LONG, 0);
        final var handler = methods.get(methodId);

        final var handle = instances.get(instancePtr.address());
        if (handler == null || handle == null) {
            setError(error, CALL_OK, 0, 0);
            ffi.variantNewNil(returnValue);
            return;
        }

        handler.invoke(handle.instance, args, argCount, returnValue, error, ffi);
    }

    private static void setError(final MemorySegment error, final int code, final int argument, final int expected) {
        if (error == MemorySegment.NULL) {
            return;
        }
        GDExtensionCallError.error(error, code);
        GDExtensionCallError.argument(error, argument);
        GDExtensionCallError.expected(error, expected);
    }

    /** Dispatches a ClassDB method call into Java. */
    @FunctionalInterface
    public interface MethodHandler {
        void invoke(
                GodotClass instance,
                MemorySegment args,
                long argCount,
                MemorySegment returnValue,
                MemorySegment error,
                GodotFFI ffi);
    }

    /** Defines the ClassDB entry and the factory for creating Java instances. */
    public record ClassDefinition(String className, String parentClassName, Supplier<? extends GodotClass> factory) {}

    private record ClassEntry(
            ClassDefinition definition,
            MemorySegment className,
            MemorySegment parentName,
            MemorySegment classUserdata) {}

    private static final class InstanceHandle implements AutoCloseable {
        private final GodotClass instance;
        private final Arena arena;
        private final MemorySegment dataPointer;
        private final long address;

        private InstanceHandle(final GodotClass instance, final Arena arena, final MemorySegment dataPointer) {
            this.instance = instance;
            this.arena = arena;
            this.dataPointer = dataPointer;
            this.address = dataPointer.address();
        }

        static InstanceHandle create(final Supplier<? extends GodotClass> factory) {
            final var instance = factory.get();
            requireNonNull(instance, "Instance factory returned null");
            final var arena = Arena.ofShared();
            final var dataPointer = arena.allocate(ValueLayout.JAVA_LONG);
            dataPointer.set(ValueLayout.JAVA_LONG, 0, dataPointer.address());
            return new InstanceHandle(instance, arena, dataPointer);
        }

        @Override
        public void close() {
            arena.close();
        }
    }
}

File: godot-java-bridge/src/main/kotlin/io/github/kingg22/godot/internal/bridge/ClassDBBridge.kt
package io.github.kingg22.godot.internal.bridge

import io.github.kingg22.godot.internal.ffm.GDExtensionCallErrorType
import io.github.kingg22.godot.internal.ffm.GDExtensionClassCreateInstance2
import io.github.kingg22.godot.internal.ffm.GDExtensionClassFreeInstance
import io.github.kingg22.godot.internal.ffm.GDExtensionClassMethodCall
import io.github.kingg22.godot.internal.ffm.GDExtensionClassToString
import io.github.kingg22.godot.internal.wrapper.GDExtensionCallError
import io.github.kingg22.godot.internal.wrapper.GDExtensionClassCreationInfo5
import io.github.kingg22.godot.internal.wrapper.GDExtensionClassMethodInfo
import java.lang.foreign.Arena
import java.lang.foreign.MemorySegment
import java.lang.foreign.ValueLayout
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Supplier

/** High-level ClassDB registration and instance dispatch.  */
class ClassDBBridge internal constructor(
    private val ffi: GodotFFI,
    private val stringNames: StringNameCache,
    private val arena: Arena,
) : AutoCloseable {
    private val classIdGenerator = AtomicLong(1)
    private val methodIdGenerator = AtomicLong(1)
    private val classes: MutableMap<Long, ClassEntry<Any>> = ConcurrentHashMap()
    private val methods: MutableMap<Long, MethodHandler> = ConcurrentHashMap()
    private val instances: MutableMap<Long, InstanceHandle<Any>> = ConcurrentHashMap()

    private val methodCallStub: MemorySegment
    private val createInstanceStub: MemorySegment
    private val freeInstanceStub: MemorySegment
    private val instanceToString: MemorySegment

    private val onMethodCall = GDExtensionClassMethodCall.Function {
            methodUserdata: MemorySegment,
            instancePtr: MemorySegment,
            args: MemorySegment,
            argCount: Long,
            returnValue: MemorySegment,
            error: MemorySegment,
        ->
        val methodId = methodUserdata.reinterpret(ValueLayout.JAVA_LONG.byteSize())
            .get(ValueLayout.JAVA_LONG, 0)
        val handler = methods[methodId]

        val handle = instances[instancePtr.address()]
        if (handler == null || handle == null) {
            GDExtensionCallError(error, GDExtensionCallErrorType.GDEXTENSION_CALL_OK, 0, 0)
            ffi.variantNewNil(returnValue)
            return@Function
        }

        handler.invoke(handle.instance, args, argCount, returnValue, error, ffi)
    }

    private val onCreateInstance = GDExtensionClassCreateInstance2.Function {
            classUserdata: MemorySegment,
            notifyPostInitialize: Byte,
        ->
        val classId =
            classUserdata.reinterpret(ValueLayout.JAVA_LONG.byteSize()).get(ValueLayout.JAVA_LONG, 0)
        val entry = classes[classId] ?: return@Function MemorySegment.NULL

        val objectPtr = ffi.classdbConstructObject2(entry.parentName)
        if (objectPtr == MemorySegment.NULL) {
            return@Function MemorySegment.NULL
        }

        val handle = InstanceHandle.create(entry.definition.factory)
        instances[handle.address] = handle
        ffi.objectSetInstance(objectPtr, entry.className, handle.dataPointer)
        return@Function objectPtr
    }

    private val onFreeInstance = GDExtensionClassFreeInstance.Function {
            classUserdata: MemorySegment,
            instancePtr: MemorySegment,
        ->
        val address = instancePtr.address()
        instances.remove(address)?.close()
    }

    init {
        this.methodCallStub = GDExtensionClassMethodCall.allocate(this.onMethodCall, arena)
        this.createInstanceStub = GDExtensionClassCreateInstance2.allocate(this.onCreateInstance, arena)
        this.freeInstanceStub = GDExtensionClassFreeInstance.allocate(this.onFreeInstance, arena)
        this.instanceToString = GDExtensionClassToString.allocate(
            GDExtensionClassToString.Function { instance: MemorySegment?, _: MemorySegment?, out: MemorySegment? ->
                val handle = instances[instance!!.address()]
                if (handle == null) {
                    System.err.println("Not found instance with address: '${instance.address()}' for toString")
                    return@Function
                }
                try {
                    val message = handle.instance.toString()
                    ffi.stringNameNewWithUtf8Chars(out!!, arena.allocateFrom(message))
                } catch (e: Exception) {
                    System.err.println("Catch exception during toString: " + e.message)
                }
            },
            arena,
        )
    }

    override fun close() {
        instances.values.forEach { obj -> obj.close() }
        instances.clear()
        classes.clear()
        methods.clear()
    }

    fun <T : Any> registerClass(definition: ClassDefinition<T>) {
        println("Registering class named : '${definition.className}' with parent: ${definition.parentClassName}")
        val classId = classIdGenerator.getAndIncrement()
        val classUserdata: MemorySegment = arena.allocate(ValueLayout.JAVA_LONG)
        classUserdata.set(ValueLayout.JAVA_LONG, 0, classId)

        val className = stringNames.get(definition.className)
        val parentName = stringNames.get(definition.parentClassName)

        val creationInfo = GDExtensionClassCreationInfo5(
            isVirtual = false,
            isAbstract = false,
            isExposed = true,
            isRuntime = true,
            createInstanceFunc = createInstanceStub,
            freeInstanceFunc = freeInstanceStub,
            classUserdata = classUserdata,
        )

        classes[classId] = ClassEntry(definition, className, parentName, classUserdata)
        ffi.classdbRegisterExtensionClass5(className, parentName, creationInfo)
    }

    fun registerMethod(className: String, methodName: String, handler: MethodHandler) {
        val methodId = methodIdGenerator.getAndIncrement()
        val methodUserdata = arena.allocate(ValueLayout.JAVA_LONG)
        methodUserdata.set(ValueLayout.JAVA_LONG, 0, methodId)
        methods[methodId] = handler

        val methodInfo = GDExtensionClassMethodInfo(
            name = stringNames.get(methodName),
            callFunc = methodCallStub,
            methodFlags = 0,
            hasReturnValue = false,
            returnValueMetadata = 0,
            argumentCount = 0,
            defaultArgumentCount = 0,
            methodUserdata = methodUserdata,
        )

        ffi.classdbRegisterExtensionClassMethod(stringNames.get(className), methodInfo)
    }

    /** Dispatches a ClassDB method call into Java.  */
    interface MethodHandler {
        fun <T> invoke(
            instance: T,
            args: MemorySegment,
            argCount: Long,
            returnValue: MemorySegment,
            error: MemorySegment,
            ffi: GodotFFI,
        )
    }

    /** Defines the ClassDB entry and the factory for creating Java instances.  */
    @JvmRecord
    data class ClassDefinition<out T>(val className: String, val parentClassName: String, val factory: Supplier<out T>)

    @JvmRecord
    private data class ClassEntry<out T>(
        val definition: ClassDefinition<T>,
        val className: MemorySegment,
        val parentName: MemorySegment,
        val classUserdata: MemorySegment,
    )

    @ConsistentCopyVisibility
    private data class InstanceHandle<out T> private constructor(
        val instance: T?,
        val arena: Arena,
        val dataPointer: MemorySegment,
    ) : AutoCloseable by arena {
        val address = dataPointer.address()
        companion object {
            fun <T> create(factory: Supplier<out T>): InstanceHandle<T> {
                val instance: T? = factory.get()
                requireNotNull(instance) { "Instance factory returned null" }
                val arena: Arena = Arena.ofShared()
                val dataPointer: MemorySegment = arena.allocate(ValueLayout.JAVA_LONG)
                dataPointer.set(ValueLayout.JAVA_LONG, 0, dataPointer.address())
                return InstanceHandle(instance, arena, dataPointer)
            }
        }
    }
}
